File: 2 - Sorting and Order Statistics/2.1 - Heap Sort/Code/priority_queue.py
"""
priority_queue.py

TODO: Implement class MinPriorityQueue

"""

from .heapify import Heapify


class MaxHeapPriorityQueue(Heapify):
    """Priority Queue Class"""

    def __init__(self, heap, tasks, queue):
        super().__init__(heap)
        self.tasks = tasks
        self.queue = queue

    def maximum(self):
        """Returns the max value of the max-heap

        Time complexity: O(1) (since we always return the first element for max-heaps)
        Space complexity: O(1)

        Raises:
            ValueError: If there are no elements in the heap

        Returns:
            int: Max priority value
        """
        if self.heap_size < 1:
            raise ValueError("Heap underflow")

        return self.heap[0]

    def extract_max(self):
        """Removes and returns the max value of the max-heap. With a max-heap, this will always be
        the element at index 0

        Time complexity: O(log n)
        Space complexity: O(n)

        Returns:
            int: Max priority value
        """
        max_elem = self.maximum()

        # Delete priority, and the task associated with it
        del self.heap[0]
        del self.tasks[0]
        del self.queue[max_elem]

        # Update heap size and fix the heap
        self.heap_size -= 1
        self.max_heapify(0)

        return max_elem

    def increase_key(self, selected_task, k):
        """Increase key operation

        Args:
            selected_task (str): Task subjected to a priority (key) change
            k (int): New key for 'selected task'

        Time complexity: O(log n)
        Space complexity: O(n)

        Raises:
            ValueError: When 'k' is smaller than the original key
        """
        old_key = next(
            (key for key, value in self.queue.items() if value == selected_task)
        )

        # Validate new key
        if k < old_key:
            raise ValueError("New key is smaller than current key")

        # Update 'queue' and 'heap' with the new priority
        self.queue = {
            k if key == old_key else key: value for key, value in self.queue.items()
        }
        self.heap[self.heap.index(old_key)] = k

        # Fix the new heap to be max-heap compliant, update queue accordingly
        i = self.heap.index(k)

        while i > 0 and self.heap[self.parent(i)] < self.heap[i]:
            self.heap[i], self.heap[self.parent(i)] = (
                self.heap[self.parent(i)],
                self.heap[i],
            )

            self.tasks[i], self.tasks[self.parent(i)] = (
                self.tasks[self.parent(i)],
                self.tasks[i],
            )

            i = self.parent(i)

        # Re-construct queue with new priorities
        self.queue = {self.heap[i]: self.tasks[i] for i in range(self.heap_size)}

    def insert(self, task, priority):
        """Insert a task into the priority queue

        Time complexity: O(log n)
        Space complexity: O(1)

        Args:
            task (str): Name of task to be inserted
            priority (int): Priority value

        """
        self.heap_size += 1
        k = priority
        priority = float("-inf")
        self.heap.append(priority)
        self.tasks.append(task)
        self.queue[priority] = task
        self.increase_key(self.tasks[-1], k)

File: 2 - Sorting and Order Statistics/2.1 - Heap Sort/Code/PriorityQueue.mjs
/**
 * PriorityQueue.mjs
 */

import Heapify from "./Heapify.mjs";

export default class MaxHeapPriorityQueue extends Heapify {
    constructor(heap, tasks, queue) {
        super(heap);
        this.tasks = tasks;
        this.queue = queue;
    }

    maximum() {
        if (this.heapSize < 1) {
            throw new Error("Heap underflow");
        }

        return this.heap[0];
    }

    extractMax() {
        let maxElem = this.maximum();

        this.heap.shift();
        this.tasks.shift();
        this.queue.delete(maxElem);

        this.heapSize -= 1;
        this.maxHeapify(0);

        return maxElem;
    }

    increaseKey(selectedTask, k) {
        let oldKey = Number.NEGATIVE_INFINITY;

        for (let [key, val] of this.queue.entries()) {
            if (val === selectedTask) {
                oldKey = key;
            }
        }

        if (k < oldKey) {
            throw new Error("New key is smaller than the current key");
        }

        this.heap[this.heap.indexOf(oldKey)] = k;

        let i = this.heap.indexOf(k);

        while (i > 0 && this.heap[this.parent(i)] < this.heap[i]) {
            let temp1 = this.heap[i];
            this.heap[i] = this.heap[this.parent(i)];
            this.heap[this.parent(i)] = temp1;

            let temp2 = this.tasks[i];
            this.tasks[i] = this.tasks[this.parent(i)];
            this.tasks[this.parent(i)] = temp2;

            i = this.parent(i);
        }
        this.queue = constructQueue(this.heap, this.tasks);
    }

    insert(task, priority) {
        this.heapSize += 1;
        let k = priority;
        priority = Number.NEGATIVE_INFINITY;
        this.heap.push(priority);
        this.tasks.push(task);
        this.queue.set(priority, task);
        this.increaseKey(this.tasks[-1], k);
    }
}

const logData = (queueObj, action) => {
    console.log();
    console.log({ action });
    console.log(`Heap: ${queueObj.heap}`);
    console.log(`Task List: ${queueObj.tasks}`);
    console.log("Queue:");

    for (var [key, value] of queueObj.queue.entries()) {
        console.log(`${key}: ${value}`);
    }
};

const constructQueue = (keys, values) => {
    let queue = new Map();

    for (var i = 0; i < keys.length; i++) {
        queue.set(keys[i], values[i]);
    }

    return queue;
};

const main = () => {
    let maxHeap = [16, 14, 10, 8, 7, 9, 3, 2, 4, 1];
    let tasks = [
        "Task1",
        "Task2",
        "Task3",
        "Task4",
        "Task5",
        "Task6",
        "Task7",
        "Task8",
        "Task9",
        "Task10",
    ];

    let queue = constructQueue(maxHeap, tasks);

    let maxPriorityQueue = new MaxHeapPriorityQueue(maxHeap, tasks, queue);
    logData(maxPriorityQueue, "Original");

    logData(maxPriorityQueue, "Max");
    console.log(`Maximum priority value: ${maxPriorityQueue.maximum()}`);

    let extractedMax = maxPriorityQueue.extractMax();
    logData(maxPriorityQueue, "Extract Max");
    console.log(
        `Highest priority ('${extractedMax}') task was removed from the queue`
    );

    maxPriorityQueue.increaseKey(tasks[8], 11);
    logData(maxPriorityQueue, "Increase Key");

    let newTask = "Urgent Matter!";
    let newPriority = 17;
    maxPriorityQueue.insert(newTask, newPriority);
    logData(maxPriorityQueue, "Insert");
    console.log(
        `Task '${newTask}' of priority '${newPriority}' was added to the queue`
    );
};

main();
